src/wmtk/PolygonMesh.cpp

```diff
@@ -0,0 +1,573 @@
+        if (!is_vertex_connectivity_valid(vid)) {
+            return false;
+        }
+    }
+
+    // Edge Connectivity
+    long n_edges = 0;
+    for (long eid = 0; eid < capacity(PrimitiveType::Edge); ++eid) {
+        // Count edges
+        if (e_flag_accessor.index_access().scalar_attribute(eid) == 0) {
+            continue;
+        } else {
+            ++n_edges;
+        }
+
+        // Check local edge validity
+        if (!is_edge_connectivity_valid(eid)) {
+            return false;
+        }
+    }
+
+    // Face Connectivity
+    long n_faces = 0;
+    for (long fid = 0; fid < capacity(PrimitiveType::Face); ++fid) {
+        // Count faces
+        if (f_flag_accessor.index_access().scalar_attribute(fid) == 0) {
+            continue;
+        } else {
+            ++n_faces;
+        }
+
+        // Check local face validity
+        if (!is_face_connectivity_valid(fid)) {
+            return false;
+        }
+    }
+
+    // Check element counts are consistent
+    if (n_halfedges != (2 * n_edges)) {
+        return false;
+    }
+    if (n_vertices != count_vertices_slow()) {
+        return false;
+    }
+    if (n_faces != count_faces_slow()) {
+        return false;
+    }
+
+    return true;
+}
+
+bool PolygonMesh::is_valid(const Tuple& tuple, ConstAccessor<long>& hash_accessor) const
+{
+    if (tuple.is_null()) return false;
+    if (tuple.m_global_cid < 0) {
+        return false;
+    }
+
+    // Check if local vid is in {0, 1}
+    if ((tuple.m_local_vid != 0) && (tuple.m_local_vid != 1)) {
+        return false;
+    }
+
+    return Mesh::is_hash_valid(tuple, hash_accessor);
+}
+
+bool PolygonMesh::is_ccw(const Tuple& tuple) const
+{
+    return (tuple.m_local_vid == 0);
+}
+
+bool PolygonMesh::is_boundary(const Tuple& tuple) const
+{
+    return is_boundary_edge(tuple);
+}
+
+bool PolygonMesh::is_boundary_vertex(const Tuple& tuple) const
+{
+    Tuple v_tuple = tuple;
+    do {
+        if (is_boundary(v_tuple)) {
+            return true;
+        }
+        v_tuple = switch_edge(switch_face(v_tuple));
+    } while (v_tuple != tuple);
+
+    return false;
+}
+
+bool PolygonMesh::is_boundary_edge(const Tuple& tuple) const
+{
+    return (is_hole_face(tuple)) || (is_hole_face(opp_halfedge(tuple)));
+}
+
+Tuple PolygonMesh::next_halfedge(const Tuple& h_tuple) const
+{
+    return switch_tuple(switch_tuple(h_tuple, PrimitiveType::Vertex), PrimitiveType::Edge);
+}
+
+Tuple PolygonMesh::prev_halfedge(const Tuple& h_tuple) const
+{
+    return switch_tuple(switch_tuple(h_tuple, PrimitiveType::Edge), PrimitiveType::Vertex);
+}
+
+Tuple PolygonMesh::opp_halfedge(const Tuple& h_tuple) const
+{
+    long h = h_tuple.m_global_cid;
+    return Tuple(
+        h_tuple.m_local_vid,
+        h_tuple.m_local_eid,
+        h_tuple.m_local_fid,
+        implicit_opp(h),
+        h_tuple.m_hash);
+}
+
+bool PolygonMesh::is_local_connectivity_valid(const Tuple& tuple, PrimitiveType type) const
```

Comment:
i know this is against the conventiosn of the other mesh libs so far / maybe you don't need to do this quite yet, but i think these valid checks don't need to be part of the polygonmesh implementation but rather belong in the debug wrappers we have for tehse classes

Reply:
(Discussed in person)